### PR TITLE
Add pickling support for WorkerInfo

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73373 [Dynamic RPC] Allow newly joined ranks to communicate with existing ranks
* #73372 [Dynamic RPC] Allow for optional world_size argument in init_rpc
* **#73371 Add pickling support for WorkerInfo**

This PRs allows for the pybinded class `WorkerInfo` to be pickled. The class is pickled into a tuple of worker_name and rank in format `(NAME, ID)`. This allows WorkerInfo to be passed as an argument for RPC calls.

Differential Revision: [D34458153](https://our.internmc.facebook.com/intern/diff/D34458153)

### CODE REVIEW

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34458153).

File: torch/csrc/distributed/rpc/init.cpp [hunk 1]
```diff
@@ -110,11 +110,26 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
...
+                os << workerInfo;
+                return os.str();
+              })
+          .def(py::pickle(
+              [](const WorkerInfo& workerInfo) { //__getstate__
+                return py::make_tuple(workerInfo.name_, workerInfo.id_);
+              },
+              [](py::tuple t) { // __setstate__
+                if (t.size() != 2) {
+                  throw std::runtime_error("Invalid state");
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp:126]: use `TORCH_CHECK`?

File: torch/csrc/distributed/rpc/init.cpp [hunk 2]
```diff
@@ -110,11 +110,26 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
...
+                }
+
+                WorkerInfo workerInfo(
+                    t[0].cast<std::string>(), t[1].cast<worker_id_t>());
+                return workerInfo;
+              }));
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp:132]: Let's add some tests for these?

**Review by mrshenli — APPROVED** [review]
LGTM

File: torch/csrc/distributed/rpc/init.cpp (hunk 1 above)
Comment (H-Huang) [reply torch/csrc/distributed/rpc/init.cpp]: Thanks! Changed it

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34458153).

File: torch/csrc/distributed/rpc/init.cpp (hunk 2 above)
Comment (H-Huang) [reply torch/csrc/distributed/rpc/init.cpp:132]: Thanks :) Added a test in rpc_test to do a simple rpc_sync call using WorkerInfo to test this.

**Pull request closed without merging**